private/tools/java/com/github/bazelbuild/rules_jvm_external/maven/MavenPublisher.java

```diff
@@ -261,6 +385,60 @@ private static String toHexS(String fmt, String algorithm, byte[] toHash) {
     }
   }
 
+  /**
+   * Attempts to download the file at the given targetUrl.
+   * Valid protocols are: http(s) & file at the moment.
+   */
+  private static CompletableFuture<Optional<String>> download(
```

Comment:
nit: could move this to a separate file `Downloaders` to keep this file a bit smaller.